Autopilot GKE documentation for Elastic Agent (#1961)
elastic/elastic-agent: docs/elastic-agent-gke-autopilot.md

Verify that you are connected to your cluster:

```bash=
>kubectl config get-contexts
CURRENT   NAME                                                                          CLUSTER                                                                       AUTHINFO                                                                      NAMESPACE
*         gke_elastic-obs-integrations-dev_us-central1_cloudnativeautopilot-cluster-1   gke_elastic-obs-integrations-dev_us-central1_cloudnativeautopilot-cluster-1   gke_elastic-obs-integrations-dev_us-central1_cloudnativeautopilot-cluster-1
```

> In case of problems to connect to your kubernetes cluster visit `Deploy` tab under `https://console.cloud.google.com/kubernetes/clusters/details/<region>/<name_cluster>/details?project=<google_project>`.

1. **Step 3: Install Kubernetes Integration in your Elastic Cloud**

Kubernetes observability is supported through both Elastic Agent installation scenarios ([managed](https://www.elastic.co/guide/en/fleet/current/install-fleet-managed-elastic-agent.html) and [standalone](https://www.elastic.co/guide/en/fleet/current/install-standalone-elastic-agent.html)).

> The Kubenretes Integration is an additional package to be installed in your Elastic Cloud and will enhance user experience for Kubenrtes Observability by installing aditonal assets in our Elastic stack.

Reviewer: ```suggestion
The Kubernetes Integration is an additional package to be installed in your Elastic Cloud and will enhance user experience for Kubernetes Observability by installing additional assets in our Elastic stack.
```

```

- Provide Elastic Cloud Credentials (In managed scenario you only neeed `FLEET_URL` and `FLEET_ENROLLMENT_TOKEN`)
  
```yaml
- name: FLEET_URL
  value: "https://123456.fleet.us-central1.gcp.cloud.es.io:443"
  # Elasticsearch API key used to enroll Elastic Agents in Fleet (https://www.elastic.co/guide/en/fleet/current/fleet-enrollment-tokens.html#fleet-enrollment-tokens)
  # If FLEET_ENROLLMENT_TOKEN is empty then KIBANA_HOST, KIBANA_FLEET_USERNAME, KIBANA_FLEET_PASSWORD are needed
  - name: FLEET_ENROLLMENT_TOKEN
  value: "WmpQbUVJVUJnX0kE6aDg0OGVtNG1TTDJqWlM2VGp"
  ```

2. [Elastic Agent Stanadlone Manifest]./manifests/elastic-agent-standalone-gke-autopilot.yaml)

- Set Container Image version of Elastic Agent

```yaml
containers:
    - name: elastic-agent-standalone
      image: docker.elastic.co/beats/elastic-agent:8.5.3
```

- Provide Elastic Cloud Credentials
  
```yaml
# This user needs the privileges required to publish events to Elasticsearch.
- name: ES_USERNAME
  value: "elastic"
# The basic authentication password used to connect to Elasticsearch
- name: ES_PASSWORD
  value: "changeme"
# The Elasticsearch host to communicate with
- name: ES_HOST

>   value: "https://<url>.kb.us-central1.gcp.cloud.es.io:9243"

Author: Yes like an example. 
Do you want to put something else more clear?

>   value: "https://<url>.kb.us-central1.gcp.cloud.es.io:9243"

Reviewer: The thing is that it said, `The Elasticsearch host to communicate with` but it is a Kibana URL, which confuses me.

>   value: "https://<url>.kb.us-central1.gcp.cloud.es.io:9243"

Author: updated

- name: NODE_NAME
```

5. **Step 5: Install Elastic Agent Manifest:**

For example:

```bash=
>kubectl apply -f elastic-agent-managed-kubernetes_gke_autopilot.yaml

namespace/elastic-agent created
configmap/agent-node-datastreams created
Warning: Autopilot increased resource requests for DaemonSet elastic-agent/elastic-agent-standalone to meet requirements. See http://g.co/gke/autopilot-resources
daemonset.apps/elastic-agent-standalone created
clusterrolebinding.rbac.authorization.k8s.io/elastic-agent-standalone created
rolebinding.rbac.authorization.k8s.io/elastic-agent-standalone created
rolebinding.rbac.authorization.k8s.io/elastic-agent-standalone-kubeadm-config created
clusterrole.rbac.authorization.k8s.io/elastic-agent-standalone created
role.rbac.authorization.k8s.io/elastic-agent-standalone created
role.rbac.authorization.k8s.io/elastic-agent-standalone-kubeadm-config created
serviceaccount/elastic-agent-standalone created
```

> Elastic Agent is installed in namespace `elastic-agent`

6. **Step 6: Install Kube-state metrics**

> Kube-state Metrics needs to be installed in the same namespace as Elastic Agent

```bash

> git clone git@github.com:kubernetes/kube-state-metrics.git

Reviewer: I think is better to recommend install a stable version from the Helm Chart, that allows to set the namespace to install.

https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-state-metrics

Author: done with latest commit

# GKE Observability with Elastic Agent

> Goal of this document is to assist users to Kubernetes Observability of Autopilot Google Kubernetes Engine Clusters (Autopilot GKE) with Elastic Agent. Monitoring, observing and alerting of Google Kubernetes Cluster made easy with following simple steps. Elastic Agent is installed as daemonset inside Google Kubernetes Cluster and will send metrics/ logs to a configured Elastic Cloud instance.

Reviewer 2: ```suggestion
The goal of this document is to assist users to Kubernetes Observability of Autopilot Google Kubernetes Engine Clusters (Autopilot GKE) with Elastic Agent. Monitoring, observing and alerting of a Google Kubernetes Cluster made easy with the following simple steps. Elastic Agent is installed as daemonset inside Google Kubernetes Cluster and will send metrics/logs to a configured Elastic Cloud instance.
```

> > Elastic Agent can be configured with multiple output setups, as part of this demo we will use Elastic Cloud

Reviewer 2: ```suggestion
> Elastic Agent can be configured with multiple output setups, as part of this guide we will use Elastic Cloud
```

  value: "https://123456.fleet.us-central1.gcp.cloud.es.io:443"
  # Elasticsearch API key used to enroll Elastic Agents in Fleet (https://www.elastic.co/guide/en/fleet/current/fleet-enrollment-tokens.html#fleet-enrollment-tokens)
  # If FLEET_ENROLLMENT_TOKEN is empty then KIBANA_HOST, KIBANA_FLEET_USERNAME, KIBANA_FLEET_PASSWORD are needed
  - name: FLEET_ENROLLMENT_TOKEN
  value: "WmpQbUVJVUJnX0kE6aDg0OGVtNG1TTDJqWlM2VGp"
  ```

2. [Elastic Agent Stanadlone Manifest](./manifests/elastic-agent-standalone_gke_autopilot_redused.yaml)

- Set Container Image version of Elastic Agent

```yaml
containers:
    - name: elastic-agent-standalone
      image: docker.elastic.co/beats/elastic-agent:8.5.3
```

- Provide Elastic Cloud Credentials
  
```yaml
# This user needs the privileges required to publish events to Elasticsearch.
- name: ES_USERNAME
  value: "elastic"
# The basic authentication password used to connect to Elasticsearch
- name: ES_PASSWORD
  value: "changeme"
# The Elasticsearch host to communicate with
- name: ES_HOST
  value: "https://<url>.kb.us-central1.gcp.cloud.es.io:9243"
- name: NODE_NAME
```

5. **Step 5: Install Elastic Agent Manifest:**

For example:

```bash=

> >kubectl apply -f elastic-agent-managed-kubernetes_gke_autopilot_redused.yaml

Reviewer 2: Is this `_redused` suffix actually meaningful at this point?

Author: Hmm you are right, not so much. I will remove them

namespace/elastic-agent created
configmap/agent-node-datastreams created
Warning: Autopilot increased resource requests for DaemonSet elastic-agent/elastic-agent-standalone to meet requirements. See http://g.co/gke/autopilot-resources
daemonset.apps/elastic-agent-standalone created
clusterrolebinding.rbac.authorization.k8s.io/elastic-agent-standalone created
rolebinding.rbac.authorization.k8s.io/elastic-agent-standalone created
rolebinding.rbac.authorization.k8s.io/elastic-agent-standalone-kubeadm-config created
clusterrole.rbac.authorization.k8s.io/elastic-agent-standalone created
role.rbac.authorization.k8s.io/elastic-agent-standalone created
role.rbac.authorization.k8s.io/elastic-agent-standalone-kubeadm-config created
serviceaccount/elastic-agent-standalone created
```

> Elastic Agent is installed in namespace `elastic-agent`

6. **Step 6: Install Kube-state metrics**

> Kube-state Metrics needs to be installed in the same namespace as Elastic Agent

```bash
git clone git@github.com:kubernetes/kube-state-metrics.git
cd kube-state-metrics/examples/standard

> sed -i -e 's/namespace: kube-system/namespace: elastic-agent/' *.yaml

Reviewer 2: 👍🏼
